homeassistant/components/fan/__init__.py

```diff
@@ -46,17 +48,36 @@
 DIRECTION_REVERSE = "reverse"
 
 ATTR_SPEED = "speed"
+ATTR_PERCENTAGE = "percentage"
 ATTR_SPEED_LIST = "speed_list"
 ATTR_OSCILLATING = "oscillating"
 ATTR_DIRECTION = "direction"
 
+# Invalid speeds do not conform to the entity model, but have crept
+# into core integrations at some point so we are temporarily
+# accommodating them in the transition to percentages.
+_INVALID_SPEED_ON = "on"
+_INVALID_SPEED_AUTO = "auto"
+_INVALID_SPEED_SMART = "smart"
+_INVALID_SPEEDS_FILTER = {_INVALID_SPEED_ON, _INVALID_SPEED_AUTO, _INVALID_SPEED_SMART}
```

Comment:
We need to handle all the modes that xiamoi_miio has too.

Reply:
Maybe we can just filter away all speeds but the four standard speeds instead of the inverse?

Reply:
Yikes

```
OPERATION_MODES_AIRPURIFIER = ["Auto", "Silent", "Favorite", "Idle"]
OPERATION_MODES_AIRPURIFIER_PRO = ["Auto", "Silent", "Favorite"]
OPERATION_MODES_AIRPURIFIER_PRO_V7 = OPERATION_MODES_AIRPURIFIER_PRO
OPERATION_MODES_AIRPURIFIER_2S = ["Auto", "Silent", "Favorite"]
OPERATION_MODES_AIRPURIFIER_3 = ["Auto", "Silent", "Favorite", "Fan"]
OPERATION_MODES_AIRPURIFIER_V3 = [
    "Auto",
    "Silent",
    "Favorite",
    "Idle",
    "Medium",
    "High",
    "Strong",
]
OPERATION_MODES_AIRFRESH = ["Auto", "Silent", "Interval", "Low", "Middle", "Strong"]
```

If we filter `Auto`, `Interval`, `Idle`, `Favorite`, `Idle` we should get an ordered list from lowest to highest for all of them

Reply:
Is there a reason we don't just filter on the valid speeds instead of the invalid?

Reply:
Yes, the goal is to get the speeds in a list from lowest to highest by removing speeds that are not valid or out of order so we can map them to percentages.

Since there are so many integration (and custom ones) that have worked around the entity model it seemed like we should try to keep them working though the transition period.

Reply:
Right, but wouldn't we achieve that by filtering in the valid speeds just the same as filtering out the invalid speeds?

Reply:
I did some renaming and added examples to the code.  Hopefully I did a better job communicating what I'm trying to achieve here with the last commit